README.md
# Nextdoor Service Registry

*nd_service_registry* is a Python module that provides a simple way to leverage
*Apache Zookeeper* as a dynamic configuration and service registry.

The goal of the package is to provide a single foundational class that can be
leveraged by any Python program for registering and monitoring services through
Zookeeper.

The current use cases are:

 * Register a server providing a service (and subsequently retrieve them)
 * Store a simple dict/key value in Zookeeper (and retreive it)
 * Execute callback methods whenever any node list or data setting changes
 * Get a temporary (but global!) lock while you do something

The main benefit of using this module is if you have several different tools
in your network that all need to interact with Zookeeper in a common way. The
most common functions are handled in this singular module allowing you to focus
on your own app development more.

## Installation

To install, run

    python setup.py install

or

    pip install nd_service_registry

## Usage

### Instantiating KazooServiceRegistry

Create a logger object

    >>> import logging
    >>> logging.basicConfig(level=logging.DEBUG)

To create your initial connection object

    >>> import nd_service_registry
    >>> nd = nd_service_registry.KazooServiceRegistry()

The KazooServiceRegistry object is a child of nd\_service\_registry that
conforms to our ServiceRegistry specs, while leveraging Kazoo as the backend.
The object handles all of your connection states - there is no need to
start/stop or monitor the connection state at all.

### Registering Data

#### Ephemeral Nodes

Registering a node in Zookeeper thats Ephemeral (disappears when the node
goes offline, or if the connection is lost) is highly useful for keeping
track of which servers are offering specific services.

    >>> nd.set_node('/production/ssh/server1:22')
    >>> nd.set_node('/production/ssh/server2:22')
    >>> nd.set_node('/production/ssh/server3:22')
    >>> nd.set_node('/production/web/server2:22',
                    data={'type': 'apache'})
    >>> nd.get('/production/ssh')
    {'children': [u'server1:22', u'server2:22', u'server3:22' ],
     'data': None,
     'path': '/production/ssh',
     'stat': ZnodeStat(czxid=27, mzxid=27, ctime=1355533229452,
                       mtime=1355533229452, version=0, cversion=5,
                       aversion=0, ephemeralOwner=0, dataLength=0,
                       numChildren=3, pzxid=45)}


To register the host as providing a particular service

    >>> nd.set_node('/services/ssh/server1:22', data={ 'foo': 'bar'})

Getting a list of servers at a path

    >>> nd.get('/services/ssh')
    {'children': {u'server1:22': {u'foo': u'bar',
                                  u'created': u'2012-12-15 00:45:03',
                                  u'pid': 10733}},
     'data': None,
     'path': '/services/ssh',
     'stat': ZnodeStat(czxid=6, mzxid=6, ctime=1355532303688,
                       mtime=1355532303688, version=0, cversion=1,
                       aversion=0, ephemeralOwner=0, dataLength=0,
                       numChildren=1, pzxid=7)}

#### Arbitrary Data

When you want to store arbitrary (but simple!) data objects in, its simple!

    >>> sr.set_data('/config/my_app', data={'state': 'enabled'})
    True
    >>> sr.get('/config/my_app')
    {'children': []
     'data': {u'created': u'2014-06-03 13:37:53',
              u'pid': 2546,
              u'state': u'enabled'},
     'path': '/config/my_app',
     'stat': ZnodeStat(czxid=76, mzxid=77, ctime=1401827873764,
                       mtime=1401827873766, version=1, cversion=0,
                       aversion=0, ephemeralOwner=0, dataLength=62,
                       numChildren=0, pzxid=76)}

### Distributed Locks

One of Zookeepers great features is using it as a global lock manager. We provide
two models for getting a lock. In one model, your lock is only active as long as
your code is running

    >>> with nd.get_lock('/foo', simultaneous=1):
    ...      <do some work>
    ...
    >>>

Another example is explicitly locking a path for some period of time, then
releasing it explicitly (eg, locking during one method, and waiting for an
entirely different method to handle the unlock)

    >>> nd.acquire_lock('/foo', simultaneous=1)
    >>> <do your work... >
    >>> nd.release_lock('/foo')

### Django

We initially wrote this code to simplify our use of Zookeeper in Django.
Included is a very simple Django utility package that makes it dead simple
to use Zookeeper inside Django for retreiving information. To use, add
the following to your *settings* module in Django:

    from nd_service_registry.contrib.django import utils as ndsr
    ndsr.SERVER = '127.0.0.1'
    ndsr.PORT = '2181'
    ndsr.TIMEOUT = 5
    ndsr.CACHEFILE = '/tmp/serviceregistry.cache'

Example usage to grab a single string and use it as a setting:
    
    MY_CRED = ndsr.get('/creds/some_cred')['data']['mycred']

Example use in your code, calling a method every time a value is updated
in Zookeeper: 

    >>> def do_something(data):
    ...     print "New server data: %s" % data
    ...
    >>> ndsr.get('/services/staging/uswest2/memcache', callback=do_something)
    New server data: { 'path': '/services/staging/uswest2/memcache',
                       'stat': ZnodeStat(czxid=8589934751, mzxid=8589934751,
                                         ctime=1354785240728,
                                         mtime=1354785240728, version=0,
                                         cversion=45, aversion=0,
                                         ephemeralOwner=0, dataLength=0,
                                         numChildren=1, pzxid=30064903926),
                       'data': None,
                       'children': { u'ec2-123-123-123-123.us-west-2:11211': {
                                       u'created': u'2013-01-08 16:51:12',
                                       u'pid': 3246
                                     }
                                   }
                       }

Warning: LC\_ALL and LANG settings
  Due to an unknown bug, if Django cannot find your LC\_ALL LOCALE settings
  (which often default to 'C'), *nd_service_registry* or *kazoo* crash and
  burn during the init phase. Its unknown why at this point, but we've found
  that its best to *unset LC\_ALL* and set *LANG=en_US:UTF-8* (or some other
  valid setting) before you start up your Django app.

  If you use Celery, set these options in */etc/default/celeryd*.

  If you use uWSGI, set them in your uWSGI config file.

  Running the Django shell

      # unset LC_ALL; LANG=en_US:UTF-8 python manage.py shell


## Connection Handling

The ServiceRegistry object tries everything that it can to make sure that
the backend Zookeeper connection is always up and running.

### Fork Behavior

In the event that your code has created an ServiceRegistry object but then
gone and forked the process (celery, as an example), we do our best to
detect this and re-create the connection, watchers and registrations.

When we detect a fork (more on that below), we re-create our Zookeeper
connection, and then re-create all Watcher and Registration objects as well.

### Fork Detection

Detecting the fork is extremely tricky... we can only really detect it when
you call the module for new data. This means that if you have created a
Watcher or Registration object, those objects will not be aware of the fork
(and thus the loss of their connection to Zookeeper) until you make another
call to them.

Because of this, I strongly recommend that if you can detect the fork from
within your application (Django signals perhaps?), you should immediately call
the *rebuild()* method on your ServiceRegistry object.

    >>> from nd_service_registry import KazooServiceRegistry
    >>> k = KazooServiceRegistry()
    >>> do_fork()
    >>> k.rebuild()

## Exceptions

The goal of this module is to be as self-contained as possible and require
as little code in your app as possible. To that end, we *almost never* raise
an Exception once the module is loaded up and connected.

We do raise a few exceptions, and each one is documented here. Whenever we
can though, we instead just *return False* as a way of indicating that we were
unable to perform your command now ... but that we will take care of it later.
Whenever we do this, we throw a WARNING log message as well.

### nd\_service\_registry.exceptions.NoConnection

Thrown if you attempt any operation that requires immediate access to the
backend Zookeeper service. Either a *set()* operation, or a *get()*
operation on a path for the first time.

Also thrown during initial connection to Zookeeper, if *lazy=False*.

(It should be noted, a *get()* will actually return the cached results even
if Zookeeper is down. This allows the service to fail temporarily in the
background but your app is still able to get the *last known* results.)

### nd\_service\_registry.exceptions.ReadOnly

If *readonly=True*, any operation that would result in a *write* will throw
this exception. Most notably, a *set()* operation will fail with this
exception if *readonly=True*.

## API Documentation

Detailed implementation details and instructions are in the individual
library files.

## Development

### Unit Tests

Running them

    $ python setup.py test
    running test
    test_creates_simple_json_outputformat (nd_service_registry.bin.ndsr.get_tests.GetTests) ... ok
    ...
    ...
    test_default_data_produces_expected_dict (nd_service_registry.funcs_tests.FuncsTests) ... ok
    test_encode_creates_dict_from_single_string (nd_service_registry.funcs_tests.FuncsTests) ... ok
    Ran 15 tests in 0.108s

    OK

### Integration Tests

Ingegration tests download and install Zookeeper in a temporary path in your
workspace.  They can be executed like this

    $ make integration
    running integration
    ...
    Make sure that the enter/exit functionality works in non-blocking ... ok
    Test that a blocking Lock works ... ok
    test_decode_converts_json_to_dict (nd_service_registry.funcs_tests.FuncsTests) ... ok

    ----------------------------------------------------------------------
    Ran 19 tests in 3.360s

    OK

## Apps using nd\_service\_registry

 * [zk_monitor](http://github.com/nextdoor/zkmonitor): Zookeeper Path Monitoring Agent
 * [zk_watcher](http://github.com/nextdoor/zkwatcher): Registers third-party (memcache/etc)
   services in Zookeeper.


CHANGELOG.md
## Version 1.0.1 (in progress)
 * Handle case where Kazoo watchers.ChildrenWatch calls _func() with a Bool.
 * Ignore Ping messages from kazoo.client logger as well.

## Version 1.0.0
 * Removed untested PyPY support from the Travis CI test framework.
 * #49: Do not set ACLs on recursively created paths.
 * #39: Completely remove get_data() calls from the children of watched paths.
 * #48: Fix compatibility with Kazoo 2.0

## Version 0.3.2
 * Bugfix: Ensure Kazoo is between 1.3 and 2.0.

## Version 0.3.1
 * Bugfix: Remove *nose* requirement from the setup.py.

## Version 0.3.0
 * #45: Prevent duplicat callback executions when data has not changed
 * #39: Reduce the number of Zookeeper get() calls made by the Watcher objects
 * #38: Ensure that an non-existent path will get a Child Watch when available
 * #37: Fix DummyWatcher behavior when no local cache data is available
 * #36: Add a get_state() method to KazooServiceRegistry
 * #35: Prevent caching of nodes where *data* and *stat* are None
 * #28: Provide normal *directory style* format for ndsr output
 * #27: Disable Rate Limiting of Zookeeper calls in the CLI *ndsr* tool
 * #25: Add DataNode object type and appropriate set_data() methods
 * #15: Implemented initial Unit and Integration tests
 * #14/#24: Add in a CLI Management Command *ndsr* for interacting with Zookeeper data
 * Testing: Integrated Travis CI testing
 * Add contrib.django.utils package for easy Django/Zookeeper integration

## Version 0.2.9
 * Fix Registration.registration again - set the *ephemeral* setting at construction time.

## Version 0.2.8
 * Fix Registration.registration - set _ephemeral = False in __init__.

## Version 0.2.7
 * #20: Fix Registration.update() method - check that type(state) is bool.

## Version 0.2.6
 * Quiet down the Kazoo logger

## Version 0.2.5
 * #18: Fix bogus log.debug() statement that caused KeyError exception.
 * #19: Make sure we call set_state any time *state* is a boolean.

## Version 0.2.4
 * Bugfix: If kazoo.Semaophore() is in canceled state, retry the lock.
 * Bugfix: release_lock() was calling _get_lock() with missing parameters
 * Bugfix: Cancel a pending lock when Lock.release() is called
 * #17: Replace time.sleep() with client.handler.sleep_func()
 * #16: Add *wait=X* option to Lock objects

## Version 0.2.3
 *  Misc documentation cleanup
 *  Added Gerrit code review docs and hooks
 *  Explicitly filter out Kazoo *PING* messages with a Logging Filter
 *  Moved *Fork Detected* messages to logging.INFO
 *  #13: Use package-level logging objects rather than class-level
 *  #12: Support for Kazoo 1.0b1 enhancement
 *  #11: Implement a Lock handler model enhancement
 *  #10: Zookeeper API throttling should be configurable
 *  #9: Exception NoNodeError can be thrown when re-connecting to Zookeeper after connection loss

## Version 0.2.2
 *  Quiet down *shutdown* function
 *  Clean up the logging object names
 *  #8: Fix funcs.save_dict() when permissions are incorrect
 *  #6: Fix *on fork* behavior


nd_service_registry/watcher.py
# Copyright 2014 Nextdoor.com, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Kazoo Zookeeper znode watch object

Copyright 2014 Nextdoor Inc."""

__author__ = 'matt@nextdoor.com (Matt Wise)'

import logging

from kazoo.recipe import watchers

from nd_service_registry import funcs


TIMEOUT = 30

log = logging.getLogger(__name__)


class Watcher(object):

    """Watches a Zookeeper path for children and data changes.

    This object provides a way to access and monitor all of the data on a
    given Zookeeper node. This includes its own node data, its children, and
    their data. It is not recursive.

    The object always maintains a local cached copy of the current state of
    the supplied path. This state can be accessed at any time by calling
    the get() method. Data will be returned in this format:

        {
            'data': { 'foo': 'bar', 'abc': 123' },
            'stat': ZnodeStat(czxid=116, mzxid=4032, ctime=1355424939217,
                    mtime=1355523495703, version=5, cversion=1912, aversion=0,
                    ephemeralOwner=0, dataLength=9, numChildren=2, pzxid=8388),
            'children': {
                'node1:22': { 'data': 'value' },
                'node2:22': { 'data': 'value2' },
            },
            'path': '/services/foo',
        }
    """

    def __init__(self, zk, path, callback=None, watch_children=True):
        """Initialize the Watcher object and begin watching a path.

        The initialization of a Watcher object automatically registers a
        data watch in Zookeeper on the path specificed. Any and all
        callbacks supplied during this initialization are executed as soon
        as the data is returned by Zookeeper.

        Optionally, a subsequent child watch is created on the children
        of the supplied path and again these callbacks are executed
        any time Zookeeper tells us that the children have changed
        in any way.

        args:
            zk: A kazoo.client.KazooClient object
            path: The path in Zookeeper to watch
            callback: A function to call when the path data changes
            wach_children: Whether or not to watch the children
        """
        # Set our local variables
        self._zk = zk
        self._path = path
        self._watch_children = watch_children

        # Get a lock handler
        self._run_lock = self._zk.handler.lock_object()

        # Create local caches of the 'data' and the 'children' data
        self._children = []
        self._data = None
        self._stat = None

        # Keep track of our Watcher objects in Kazoo with local references
        self._current_data_watch = None
        self._current_children_watch = None

        # Array to hold any callback functions we're supposed to notify when
        # anything changes on this path
        self._callbacks = []
        if callback:
            self._callbacks.append(callback)

        # Quick copy of the data last used when executing callbacks -- used
        # to prevent duplicate callback executions due to multiple watches.
        self._last_callback_executed_with = None

        # if self._state is False, then even on a data change, our callbacks
        # do not run.
        self._state = True

        # Start up our DataWatch. This can be started on any path regardless of
        # whether it exists or not. ChildrenWatch objects though must be
        # started on only existing-paths -- so we do not create that object
        # here, but instead do it from with the self._update() method.
        log.debug('[%s] Registering DataWatch (%s)' %
                  (self._path, self._current_data_watch))
        self._current_data_watch = watchers.DataWatch(
            client=self._zk,
            path=self._path,
            func=self._update)

    def get(self):
        """Returns local data/children in specific dict format"""
        ret = {}
        ret['stat'] = self._stat
        ret['path'] = self._path
        ret['data'] = self._data
        ret['children'] = self._children
        return ret

    def stop(self):
        """Stops watching the path."""
        # Stop the main run() method
        self._state = False

    def start(self):
        """Starts watching the path."""
        # Stop the main run() method
        self._state = True

    def state(self):
        """Returns self._state"""
        return self._state

    def add_callback(self, callback):
        """Add a callback when watch is updated."""
        for existing_callback in self._callbacks:
            if callback == existing_callback:
                log.debug('[%s] Callback [%s] already exists. Not triggering '
                          'again.' % (self._path, callback))
                return

        self._callbacks.append(callback)
        callback(self.get())

    def _update(self, data, stat):
        """Function executed by Kazoo upon data/stat changes for a path.

        This function is registered during the __init__ process of this
        Object with Kazoo. Upon the initial registration, and any subsequent
        changes to the 'data' or 'stat' of a path, this function is called
        again.

        This function is responsible for updating the local Watcher object
        state (its path data, stat data, etc) and triggering any callbacks
        that have been registered with this Watcher.

        args:
            data: The 'data' stored in Zookeeper for this path
            stat: The 'stat' data for this path
        """
        self._data = funcs.decode(data)
        self._stat = stat
        log.debug('[%s] Data change detected: %s, Stat: %s' %
                  (self._path, self._data, self._stat))

        # ChildrenWatches can only be applied to already existing paths
        # (Kazoo throws a NoNodeError otherwise). To prevent this NoNodeError
        # from being thrown, we only register an additional ChildrenWatch
        # in the event that 'stat' was not None.
        if self._watch_children and stat and not self._current_children_watch:
            log.debug('[%s] Registering ChildrenWatch' % self._path)
            self._current_children_watch = watchers.ChildrenWatch(
                client=self._zk,
                path=self._path,
                func=self._update_children)

        # Lastly, execute our callbacks
        self._execute_callbacks()

    def _update_children(self, children):
        """Function executed by Kazoo upon children changes for a path.

        This function is registered by the _update() method during and is
        executed by Kazoo upon any children changes to a path. Its responsible
        for updating the local list of children for the path, and executing
        the appropriate callbacks.

        args:
            children: The list of children returned by Kazoo
        """
        try:
            self._children = sorted(children)
        except TypeError:
            log.warning(
                '[%s] Invalid value passed to us by Kazoo: %s '
                '[see Kazoo issue #255]' %
                (self._path, children))
            return

        log.debug('[%s] Children change detected: %s' %
                  (self._path, self._children))
        self._execute_callbacks()

    def _execute_callbacks(self):
        """Runs any callbacks that were passed to us for a given path.

        Args:
            path: A string value of the 'path' that has been updated. This
                  triggers the callbacks registered for that path only.
        """

        # If the current data and the last-execution data are the same, then
        # assume our callback notification was bogus and don't run.
        if self._last_callback_executed_with == self.get():
            log.debug('[%s] Last callback data matches current data, not '
                      'executing callbacks again.' % self._path)
            return

        log.debug('[%s] execute_callbacks triggered' % self._path)
        if not self.state():
            log.debug('[%s] self.state() is False - not executing callbacks.'
                      % self._path)
            return

        for callback in self._callbacks:
            log.debug('[%s] Executing callback %s' % (self._path, callback))
            callback(self.get())

        # Store a "last called with" variable that we can check to prevent
        # unnecessary extra callback executions.
        self._last_callback_executed_with = self.get()


class DummyWatcher(Watcher):

    """Provides a Watcher-interface, without any actual Zookeeper connection.

    This object can store and return data just like a Watcher object, but
    has no connection at all to Zookeeper or any actual watches. This object
    is meant to be used in the event that Zookeeper is down and we still
    want to be able to return valid data.
    """

    def __init__(self, path, data, callback=None):
        # Set our local variables
        self._path = path
        self._data = data['data']
        self._stat = data['stat']
        self._children = data['children']

        # Array to hold any callback functions we're supposed to notify when
        # anything changes on this path
        self._callbacks = []
        if callback:
            self.add_callback(callback)

    def stop(self):
        """Stops watching the path (not applicable here)"""
        return

    def start(self):
        """Starts watching the path (not applicable here)"""
        return

    def state(self):
        """Returns state of the object (not applicable here)"""
        return True
